pkg/deviceplugin/plugin.go

```diff
@@ -0,0 +1,133 @@
+package deviceplugin
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1beta1"
+
+	"github.com/kubevirt/macvtap-cni/pkg/util"
+)
+
+const (
+	tapPath         = "/dev/tap"
+	suffix          = "Mvp"
+	defaultCapacity = 100
+)
+
+type MacvtapDevicePlugin struct {
+	Name        string
+	Master      string
+	Mode        string
+	Capacity    int
+	stopWatcher chan struct{}
+}
+
+func NewMacvtapDevicePlugin(name string, master string, mode string, capacity int) *MacvtapDevicePlugin {
+	return &MacvtapDevicePlugin{
+		Name:        name,
+		Master:      master,
+		Mode:        mode,
+		Capacity:    capacity,
+		stopWatcher: make(chan struct{}),
+	}
+}
+
+func (mdp *MacvtapDevicePlugin) generateMacvtapDevices() []*pluginapi.Device {
+	var macvtapDevs []*pluginapi.Device
+
+	var capacity = mdp.Capacity
+	if capacity <= 0 {
+		capacity = defaultCapacity
+	}
+
+	for i := 0; i < capacity; i++ {
+		name := fmt.Sprint(mdp.Name, suffix, i)
+		macvtapDevs = append(macvtapDevs, &pluginapi.Device{
+			ID:     name,
+			Health: pluginapi.Healthy,
+		})
+	}
+	return macvtapDevs
+}
+
+func masterExists(master string) bool {
+	if err := util.LinkExists(master); err != nil {
+		glog.Warningf("Master %s not found: %v", master, err)
+		return false
+	}
+	return true
+}
+
+func (mdp *MacvtapDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	masterDevs := mdp.generateMacvtapDevices()
+	noMasterDevs := make([]*pluginapi.Device, 0)
+	emitResponse := func(masterExists bool) {
+		if masterExists {
+			glog.V(3).Info("Master exists, sending ListAndWatch response with available devices")
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: masterDevs})
+		} else {
+			glog.V(3).Info("Master does not exist, sending ListAndWatch response with no devices")
+			s.Send(&pluginapi.ListAndWatchResponse{Devices: noMasterDevs})
+		}
+	}
+
+	didMasterExist := false
```

Comment:
Could you please add an inline comment documenting what are you doing in the following block?

Reply:
Added.